Add Neighborhood Attention Transformer (NAT) and Dilated NAT (DiNAT) models

alihassanijr opened the pull request:
# What does this PR do?

This PR adds [NAT](https://github.com/SHI-Labs/Neighborhood-Attention-Transformer/blob/main/NAT.md) and [DiNAT](https://github.com/SHI-Labs/Neighborhood-Attention-Transformer/blob/main/DiNAT.md) and their dependencies.

## Dependencies
- [NATTEN](https://github.com/SHI-Labs/NATTEN/) is the only new requirement. The models themselves are mostly in the same style as most `timm` models. They just require NATTEN to get the sliding window attention.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?
  - Yes, mostly boilerplate from similar models.


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten @NielsRogge @amyeroberts @sgugger

alihassanijr committed:
Add DiNAT

alihassanijr committed:
Adds DiNAT + tests

alihassanijr committed:
Minor fixes

alihassanijr committed:
Added HF model

alihassanijr commented:
Tests are failing right now because I still haven't figured out how to add the one dependency correctly so that circleci picks it up.

alihassanijr committed:
Add natten to dependencies.

alihassanijr committed:
Cleanup

alihassanijr committed:
Minor fixup

alihassanijr committed:
Reformat

alihassanijr committed:
Optional NATTEN import.

alihassanijr committed:
Reformat & add doc to _toctree

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20219). All of your documentation changes will be reflected on that endpoint.

alihassanijr commented:
I think I'm basically done (still have to push out one reformat commit to get quality tests to pass).
### What's left?
- Adding ci test
  - Right now the unittests related to the new model will be skipped, because the model relies on NATTEN. 
  - I'm unsure whether I should add it as a dependency to `torch_job` directly or not. Thoughts?

alihassanijr committed:
Reformat (finally)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20219). All of your documentation changes will be reflected on that endpoint.

alihassanijr committed:
Dummy objects for DiNAT

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20219). All of your documentation changes will be reflected on that endpoint.

alihassanijr committed:
Add NAT + minor changes

Adds NAT as its own independent model + docs, tests
Adds NATTEN to ext deps to ensure ci picks it up.

alihassanijr committed:
Remove natten from `all` and `dev-torch` deps, add manual pip install to ci tests

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20219). All of your documentation changes will be reflected on that endpoint.

alihassanijr committed:
Minor fixes.

alihassanijr committed:
Fix READMEs.

NielsRogge commented on .circleci/create_circleci_config.py:
```diff
@@ -154,6 +154,7 @@ def job_name(self):
         "sudo apt-get -y update && sudo apt-get install -y libsndfile1-dev espeak-ng time",
         "pip install --upgrade pip",
         "pip install .[sklearn,torch,testing,sentencepiece,torch-speech,vision,timm]",
+        "pip install natten",
```
I'll ping @LysandreJik and @sgugger here regarding this extra dependency. 

Ideally we don't add additional dependencies for models, for instance Deformable DETR also relies on a custom kernel and it's just a folder in the [model folder](https://github.com/huggingface/transformers/tree/main/src/transformers/models/deformable_detr).

NielsRogge commented on .circleci/create_circleci_config.py:
```diff
@@ -189,6 +190,7 @@ def job_name(self):
         "sudo apt-get -y update && sudo apt-get install -y libsndfile1-dev espeak-ng",
         "pip install --upgrade pip",
         "pip install .[sklearn,torch,testing,sentencepiece,torch-speech,vision,timm]",
+        "pip install natten",
```
Ideally we can remove this

alihassanijr replied:
Thank you, please refer to my comment below.

NielsRogge commented on README.md:
```diff
@@ -295,6 +295,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[DeiT](https://huggingface.co/docs/transformers/model_doc/deit)** (from Facebook) released with the paper [Training data-efficient image transformers & distillation through attention](https://arxiv.org/abs/2012.12877) by Hugo Touvron, Matthieu Cord, Matthijs Douze, Francisco Massa, Alexandre Sablayrolles, Hervé Jégou.
 1. **[DETR](https://huggingface.co/docs/transformers/model_doc/detr)** (from Facebook) released with the paper [End-to-End Object Detection with Transformers](https://arxiv.org/abs/2005.12872) by Nicolas Carion, Francisco Massa, Gabriel Synnaeve, Nicolas Usunier, Alexander Kirillov, Sergey Zagoruyko.
 1. **[DialoGPT](https://huggingface.co/docs/transformers/model_doc/dialogpt)** (from Microsoft Research) released with the paper [DialoGPT: Large-Scale Generative Pre-training for Conversational Response Generation](https://arxiv.org/abs/1911.00536) by Yizhe Zhang, Siqi Sun, Michel Galley, Yen-Chun Chen, Chris Brockett, Xiang Gao, Jianfeng Gao, Jingjing Liu, Bill Dolan.
+1. **[DiNAT](https://huggingface.co/docs/transformers/model_doc/dinat)** (from SHI Labs) released with the paper [Dilated Neighborhood Attention Transformer](https://arxiv.org/abs/2209.15001) by Ali Hassani and Humphrey Shi.
```
```suggestion
1. **[DiNAT](https://huggingface.co/docs/transformers/main/model_doc/dinat)** (from SHI Labs) released with the paper [Dilated Neighborhood Attention Transformer](https://arxiv.org/abs/2209.15001) by Ali Hassani and Humphrey Shi.
```
You need to run `make fix-copies` when applying this change to update the other READMEs

alihassanijr replied:
Done. I'm assuming this was referencing both NAT and DiNAT?

NielsRogge commented on docs/source/en/model_doc/dinat.mdx:
```diff
@@ -0,0 +1,75 @@
+## Overview
+
+DiNAT was proposed in [Dilated Neighborhood Attention Transformer](https://arxiv.org/abs/2209.15001)
+by Ali Hassani and Humphrey Shi.
```
Usually we add a one-line summary sentence after this

alihassanijr replied:
Done.
https://github.com/huggingface/transformers/blob/f24611ce3eab84d8af62e9cf0672c64206dbf5e0/docs/source/en/model_doc/dinat.mdx?plain=1#L20-L21

NielsRogge commented on docs/source/en/model_doc/dinat.mdx:
```diff
@@ -0,0 +1,75 @@
+and ranks second on Cityscapes (84.5 mIoU) (no extra data). *
+
+Tips:
+- One can use the [`AutoFeatureExtractor`] API to prepare images for the model.
+- DiNAT can be used as a *backbone*. When `output_hidden_states = True`,
+it will output both `hidden_states` and `reshaped_hidden_states`. The `reshaped_hidden_states` have a shape of `(batch, num_channels, height, width)` rather than `(batch_size, height, width, num_channels)`.
```
Is this also true for this model? Seems like it was copied from Swin's docs

alihassanijr replied:
Yes it is, but modified to reflect the change in hidden state shapes.

NielsRogge commented on docs/source/en/model_doc/dinat.mdx:
```diff
@@ -0,0 +1,75 @@
+[[autodoc]] DiNATConfig
+
+
+## DiNATModel
```
nit: it might be more user-friendly to name the model `DinatModel` instead of `DiNATModel` as it will be easier for people to type ;)

NielsRogge commented on docs/source/en/model_doc/nat.mdx:
```diff
@@ -0,0 +1,71 @@
+## Overview
+
+NAT was proposed in [Neighborhood Attention Transformer](https://arxiv.org/abs/2204.07143)
+by Ali Hassani, Steven Walton, Jiachen Li, Shen Li, and Humphrey Shi.
```
Same here, please add a one-sentence summary

alihassanijr replied:
https://github.com/huggingface/transformers/blob/f24611ce3eab84d8af62e9cf0672c64206dbf5e0/docs/source/en/model_doc/nat.mdx?plain=1#L20-L20

NielsRogge commented on setup.py:
```diff
@@ -125,6 +125,7 @@
     "jieba",
     "kenlm",
     "nltk",
+    "natten>=0.14.4",
```
Same here, please don't add this dependency

alihassanijr replied:
Please refer to my comment below.

NielsRogge reviewed (commented):
(removed)

NielsRogge reviewed (changes requested):
This PR seems already in a very clean state, wow! Amazing work.

I just think we'll need to move the custom kernel to a folder similar to [Deformable DETR](https://github.com/huggingface/transformers/tree/main/src/transformers/models/deformable_detr), rather than making it an additional dependency.

cc'ing @LysandreJik and @sgugger here for confirmation.

sgugger commented:
In this instance, I'd actually prefer to rely on the extra dep (as long as it's properly set up as a soft dependency, which seems to be the case in the PR). We don't know how to maintain CUDA kernels anyway, so support will be a lot better if it's done elsewhere.

alihassanijr commented:
Hi @NielsRogge @sgugger 
Actually we're happy to do it either way, but the reason we packaged NATTEN as a pip package in the first place is to make installation a lot easier, especially since we plan to upgrade it frequently.
Unlike Deformable Attention, it doesn't come with a fixed set of kernels. There's still a lot of improvements that we've planned ahead to NATTEN.

And yes, as @sgugger stated, it would work as a soft dependency; even imports aren't broken. But there's dummy calls to the package in case it's not available, that will raise an error only when the forward functions are called.

As for the torch tests, I only did those as a suggestion. I would personally recommend having a separate test for these models in general so that it doesn't get in the way. Additionally, knowing the torch build beforehand is better, since that way we can just specify a wheel URL and have it just install a lot faster.

I'll make the changes to the docs and run fix-copies again (I must have forgotten to do that in the last commit).

And yes, both models were cloned off of Swin; the architectures are somewhat similar.
The difference here is that there's a convolutional tokenizer and downsampler replacing the patch+embed and patch merging layers; and we like to keep tensors in shape B H W C, since NATTEN also expects the height and width axes to be unrolled, like a Conv2d.

alihassanijr committed:
Requested changes to docs + minor fixes.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20219). All of your documentation changes will be reflected on that endpoint.

alihassanijr commented:
Actually, I just noticed `transformers` doesn't come with wheels, right?
My previous statement about wheel sizes is irrelevant in that case.

However, I would shift more towards @sgugger 's point of view, since loading torch extensions at runtime becomes less and less reliable as extensions grow, and NATTEN already has twice the number of kernels compared to MSDeformAttn (excluding the templating that goes on in NATTEN). This would have the users wait up to 5 minutes before being able to use these models, and would affect reproducibility.

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+
+logger = logging.get_logger(__name__)
+
+# General docstring
+_CONFIG_FOR_DOC = "DiNATConfig"
+_FEAT_EXTRACTOR_FOR_DOC = "AutoFeatureExtractor"
+
+# Base docstring
+_CHECKPOINT_FOR_DOC = "shi-labs/dinat-mini-in1k-224"
+_EXPECTED_OUTPUT_SHAPE = [1, 7, 7, 512]
+
+# Image classification docstring
+_IMAGE_CLASS_CHECKPOINT = "shi-labs/dinat-mini-in1k-224"
+_IMAGE_CLASS_EXPECTED_OUTPUT = "tabby, tabby cat"
```
Feel free to also add NAT and DINAT to the documentation tests file: https://github.com/huggingface/transformers/blob/main/utils/documentation_tests.txt. This will make sure the doc tests are run daily

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+        num_channels, hidden_size = config.num_channels, config.embed_dim
+        self.num_channels = num_channels
+
+        assert patch_size == 4, "DiNAT only supports patch size of 4 at the moment."
```
Please replace assertions by `if ... else raise ValueError` patterns :)

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+        return embeddings
+
+
+class DiNATTokenizer(nn.Module):
```
```suggestion
class DiNATEmbeddings(nn.Module):
```
(nit) Cool name, although we're using the term Embeddings everywhere in the code base.. so could you use that name here as well?

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+        self.kernel_size = kernel_size
+        self.dilation = dilation
+
+        self.rpb = nn.Parameter(torch.zeros(num_heads, (2 * self.kernel_size - 1), (2 * self.kernel_size - 1)))
```
Could you add a comment to explain what "rpb" stands for?

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/nat/configuration_nat.py:
```diff
@@ -0,0 +1,136 @@
+        encoder_stride (`int`, `optional`, defaults to 32):
+            Factor to increase the spatial resolution by in the decoder head for masked image modeling.
+
+        Example:
```
This needs to be shift to the left here

alihassanijr replied:
Done. Does it look right now?

NielsRogge commented on src/transformers/models/dinat/configuration_dinat.py:
```diff
@@ -0,0 +1,140 @@
+        encoder_stride (`int`, `optional`, defaults to 32):
+            Factor to increase the spatial resolution by in the decoder head for masked image modeling.
+
+        Example:
```
This needs to be shift to the left

alihassanijr replied:
Done. Does it look right now?

NielsRogge commented on src/transformers/models/nat/modeling_nat.py:
```diff
@@ -0,0 +1,829 @@
+        return embeddings
+
+
+class NATTokenizer(nn.Module):
```
```suggestion
class NATPatchEmbeddings(nn.Module):
```
Same comment here

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/nat/modeling_nat.py:
```diff
@@ -0,0 +1,829 @@
+        num_channels, hidden_size = config.num_channels, config.embed_dim
+        self.num_channels = num_channels
+
+        assert patch_size == 4, "NAT only supports patch size of 4 at the moment."
```
Same comment here, please no assertions

alihassanijr replied:
Done.

NielsRogge commented on tests/models/dinat/test_modeling_dinat.py:
```diff
@@ -0,0 +1,347 @@
+if is_vision_available():
+    from PIL import Image
+
+    from transformers import AutoFeatureExtractor
```
I think we can update this to leverage `AutoImageProcessor` now, cc'ing @amyeroberts here

(for context, we're deprecating the FeatureExtractor classes in favor of ImageProcessor)

alihassanijr replied:
Done. Replaced in tests, docstrings, and docs.

NielsRogge commented on tests/models/nat/test_modeling_nat.py:
```diff
@@ -0,0 +1,344 @@
+if is_vision_available():
+    from PIL import Image
+
+    from transformers import AutoFeatureExtractor
```
Same comment here

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+
+
+@dataclass
+class DiNATEncoderOutput(ModelOutput):
```
```suggestion
# Copied from transformers.models.nat.modeling_nat.NATEncoderOutput with NAT->DiNAT
class DiNATEncoderOutput(ModelOutput):
```
If DiNAT is an improvement upon NAT, would it be possible to add Copied from statements everywhere where it's 
possible? This way, both modeling files keep being consistent.

NielsRogge replied:
Also note that you can also add "Copied from" statements on methods rather than classes (so in case for instance the init method can be copied but the forward not, you can add # Copied from transformers.models.nat.modeling_nat.NATEncoder.__init__ with NAT->DiNAT

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+
+
+@dataclass
+class DiNATModelOutput(ModelOutput):
```
Please add Copied from here

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+
+
+@dataclass
+class DiNATImageClassifierOutput(ModelOutput):
```
Please add Copied from here and everywhere below where it's possible

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/dinat/configuration_dinat.py:
```diff
@@ -0,0 +1,140 @@
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+            The epsilon used by the layer normalization layers.
+        encoder_stride (`int`, `optional`, defaults to 32):
+            Factor to increase the spatial resolution by in the decoder head for masked image modeling.
```
This argument can be removed, unless DiNAT supports masked image modeling

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/nat/configuration_nat.py:
```diff
@@ -0,0 +1,136 @@
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+            The epsilon used by the layer normalization layers.
+        encoder_stride (`int`, `optional`, defaults to 32):
+            Factor to increase the spatial resolution by in the decoder head for masked image modeling.
```
This argument can be removed, it's for masked image modeling only

alihassanijr replied:
Done.

NielsRogge commented on src/transformers/models/nat/modeling_nat.py:
```diff
@@ -0,0 +1,829 @@
+        return "p={}".format(self.drop_prob)
+
+
+class NeighborhoodAttention(nn.Module):
```
```suggestion
class NATNeighborhoodAttention(nn.Module):
```

NielsRogge reviewed (approved):
Wow, this PR is incredibly clean. Thanks for your work.

Main comments are about:
* adding Copied from statements wherever possible
* potentially replacing FeatureExtractor by ImageProcessor (cc @amyeroberts)

sgugger commented on .circleci/create_circleci_config.py:
```diff
@@ -154,6 +154,7 @@ def job_name(self):
         "sudo apt-get -y update && sudo apt-get install -y libsndfile1-dev espeak-ng time",
         "pip install --upgrade pip",
         "pip install .[sklearn,torch,testing,sentencepiece,torch-speech,vision,timm]",
+        "pip install natten",
```
Let's remove this from here. After internal dicussion, we will test this model with LayoutLM and make this job the test for all "exotic" models that require additional dependencies. So please add it to `layoutlm_job` below :-)

alihassanijr replied:
Done.

sgugger commented on docs/source/en/model_doc/dinat.mdx:
```diff
@@ -0,0 +1,78 @@
+and ranks second on Cityscapes (84.5 mIoU) (no extra data). *
+
+Tips:
+- One can use the [`AutoFeatureExtractor`] API to prepare images for the model.
```
```suggestion
- One can use the [`AutoImageProcessor`] API to prepare images for the model.
```
We're in the midst of deprecating the feature extractors, so let's focus all doc on image processor!

alihassanijr replied:
Done.

sgugger commented on docs/source/en/model_doc/dinat.mdx:
```diff
@@ -0,0 +1,78 @@
+The original code can be found [here](https://github.com/SHI-Labs/Neighborhood-Attention-Transformer).
+
+
+## DiNATConfig
```
I know the library is not very consistent this way but we had several complains of users that our models were too hard to find when using cased prefixes like this. So let's do `DinatConfig`, `DinatModel` etc. (like `BertModel`, `BertConfig` or `RobertaModel`, `RobertaConfig`.

alihassanijr replied:
Done.

sgugger commented on docs/source/en/model_doc/nat.mdx:
```diff
@@ -0,0 +1,73 @@
+The original code can be found [here](https://github.com/SHI-Labs/Neighborhood-Attention-Transformer).
+
+
+## NATConfig
```
Same comment as above: `NatConfig`, `NatModel` etc. please.

alihassanijr replied:
Done.

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -343,6 +345,7 @@
         ("deit", "DeiT"),
         ("detr", "DETR"),
         ("dialogpt", "DialoGPT"),
+        ("dinat", "DiNAT"),
```
Those here should be left with proper casing however!

alihassanijr replied:
Done.

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -105,6 +106,7 @@
         ("mpnet", "MPNetConfig"),
         ("mt5", "MT5Config"),
         ("mvp", "MvpConfig"),
+        ("nat", "NATConfig"),
```
Can you also add the config archive maps in the constant below?

alihassanijr replied:
Done.

sgugger commented on src/transformers/models/dinat/configuration_dinat.py:
```diff
@@ -0,0 +1,140 @@
+            The number of input channels.
+        embed_dim (`int`, *optional*, defaults to 64):
+            Dimensionality of patch embedding.
+        depths (`list(int)`, *optional*, defaults to [2, 2, 6, 2]):
+            Number of layers in each level of the encoder.
+        num_heads (`list(int)`, *optional*, defaults to [3, 6, 12, 24]):
+            Number of attention heads in each layer of the Transformer encoder.
```
```suggestion
        depths (`List[int]`, *optional*, defaults to `[2, 2, 6, 2]`):
            Number of layers in each level of the encoder.
        num_heads (`List[int]`, *optional*, defaults to `[3, 6, 12, 24]`):
            Number of attention heads in each layer of the Transformer encoder.
```
Please use the typing standard (so `List[int]`) and put the list default values in code blocks (also applies to `dilations` below.

alihassanijr replied:
Done.

sgugger reviewed (commented):
Nice, thanks a lot for adding those two models!
My main comment is around naming. As I've explained in a comment below, users prefer model class names to be non-cased for easier discoverability, so we don't add new models with special casings. Therefore, could you rename to `DinatConfig`, `DinatModel`, `NatConfig`, `NatModel` etc. ? Thanks!

alihassanijr committed:
Requested changes.

alihassanijr commented on src/transformers/models/dinat/modeling_dinat.py:
```diff
@@ -0,0 +1,836 @@
+class DiNATForImageClassification(DiNATPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
```
Done.

alihassanijr commented on src/transformers/models/nat/configuration_nat.py:
```diff
@@ -0,0 +1,136 @@
+            The number of input channels.
+        embed_dim (`int`, *optional*, defaults to 64):
+            Dimensionality of patch embedding.
+        depths (`list(int)`, *optional*, defaults to [2, 2, 6, 2]):
```
Done.

alihassanijr commented on src/transformers/models/nat/modeling_nat.py:
```diff
@@ -0,0 +1,829 @@
+)
+class NATModel(NATPreTrainedModel):
+    def __init__(self, config, add_pooling_layer=True):
+        super().__init__(config)
```
Done.

alihassanijr commented on tests/models/nat/test_modeling_nat.py:
```diff
@@ -0,0 +1,344 @@
+        )
+        if is_torch_available()
+        else ()
+    )
```
Done.

alihassanijr commented on src/transformers/models/nat/modeling_nat.py:
```diff
@@ -0,0 +1,829 @@
+)
+class NATForImageClassification(NATPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
```
Done.

alihassanijr commented on .circleci/create_circleci_config.py:
```diff
@@ -310,6 +308,7 @@ def job_name(self):
         "pip install pytesseract",
+        "pip install natten",
     ],
     tests_to_run="tests/models/*layoutlmv*",
```
@sgugger Just one last question. Do you prefer if the tests related to NAT/DiNAT are added under a new layoutlmv directory, or do I just manually include the two subdirectories (`tests/models/*nat`) to `tests_to_run`?

sgugger replied:
You should manually add the two subdirectories here, please!

alihassanijr committed:
Add NAT/DiNAT tests to layoutlm_job

alihassanijr committed:
Correction to Dinat doc.

sgugger commented on tests/models/dinat/test_modeling_dinat.py:
```diff
@@ -0,0 +1,340 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
```suggestion
        config, pixel_values, labels = config_and_inputs
```

sgugger reviewed (commented):
A couple more nits, and some of the previous comments haven't been fully addressed. Tried to flag everything below.
Thanks!

alihassanijr committed:
Requested changes.

alihassanijr commented:
@sgugger All done.
And sorry I didn't go with directly applying your changes, some of them needed a few more replacements so I figured I'd make the changes and commit all of them at once.

sgugger commented:
Thanks! You'll need a rebase on main to get the tests to pass as TensorFlow new release breaks the CI

sgugger reviewed (approved):
Thanks for all your work on those models!

alihassanijr commented:
Thanks for the reviews and feedback @sgugger @NielsRogge @amyeroberts .
Looking forward to contributing more in the future.
